tutorials/examples/mRNA_design/MFE_calculator.py

```diff
@@ -0,0 +1,559 @@
+        u0 = U[i, j1]
+        u1 = V[i, j1]
+        umin = min(u0, u1)
+        kmin = -1
+
+        # Hypothesise that j pairs with k and [i,k-1] is optimally folded.
+        for k in range(i + 1, j):
+            u0 = U[i, k - 1]
+            u1 = V[i, k - 1]
+            u2 = V[k, j]
+            if (u := min(u0, u1) + u2) < umin:
+                umin = u
+                kmin = k
+
+        return umin, kmin
+
+
+class RNAFolder(RNAFolderBase2):
+    def solve(self, rna: str):
+        # Hold the working arrays and RNA string in an object.
+        soln = RNASolution(rna)
+        N = soln.N
+        K, U, V = soln.get_arrays()
+        M = soln.M
+
+        # Subsequences up to smallest loop size contain no pairs
+        for p in range(0, self.lmin + 1):
+            for i in range(N - p):
+                j = i + p
+                V[i, j] = math.inf
+                K[i, j] = -1
+                M[i, j] = -1
+
+        for p in range(self.lmin + 1, N):
+            # in pass p, we determine open and closed free energy for
+            # subsequences of length p. (i,j) moves diagonally down and
+            # right in i,j-space with j - i == p.
+            for i in range(N - p):
+                j = i + p
+
+                # Compute the energy with (i,j) paired if admissible.
+                vmin, mode = self.closed_energy(soln, i, j)
+
+                # Compute the energy with (i,j) unpaired
+                umin, k = self.open_energy(soln, i, j)
+
+                # Store scores for structure [i,j]
+                K[i, j] = k
+                M[i, j] = mode
+                U[i, j] = umin
+                V[i, j] = vmin
+
+        return soln
+
+
+# COUNT_PSEUDO_FE = dict(base_pair_AU=-1, base_pair_CG=-1, base_pair_GU=-1)
+
+
+# def print_workings(s: RNASolution):
+#     """Tabulate the scores in the solution"""
+#     rna = s.rna
+#     N = len(rna)
+#     print("U, K:")
+#     print("     " + "".join(f"{j:5d} " for j in range(N)))
+#     print("     " + " -----" * N)
+#     for i in range(N):
+#         print(f"{i:3d} :", end="")
+#         for j in range(N):
+#             if j < i:
+#                 u = s.getU(j, i)
+#                 print("    . " if u == 0.0 else f"{u:6.1f}", end="")
+#             elif j > i:
+#                 k = s.getK(i, j)
+#                 print("    . " if k < 0 else f"{k:5d} ", end="")
+#             else:
+#                 print(f"{'=' + rna[i]:>5s} ", end="")
+#         print()
+
+#     print("V, M:")
+#     for i in range(N):
+#         print(f"{i:3d} :", end="")
+#         for j in range(N):
+#             if j < i:
+#                 v = s.getV(j, i)
+#                 print("    . " if math.isinf(v) else f"{v:6.1f}", end="")
+#             elif j > i:
+#                 m = s.getM(i, j)
+#                 print("    . " if m < 0 else f"{m:5d} ", end="")
+#             else:
+#                 print(f"{'=' + rna[i]:>5s} ", end="")
+#         print()
+
+
+# def print_ruler(rna: str):
```

Comment:
You could put all of this stuff into the classes with an optional `verbose` flag instead of leaving the code commented out here.